src/Updater/Updater.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+namespace my127\Workspace\Updater;
+
+use Error;
+use my127\Workspace\Updater\Exception\NoUpdateAvailableException;
+use my127\Workspace\Updater\Exception\NoVersionDeterminedException;
+use Phar;
+use RuntimeException;
+use Throwable;
+
+class Updater
+{
+    public const CODE_ERR_FETCHING_RELEASES = 100;
+    public const CODE_NO_RELEASES = 101;
+    public const CODE_ERR_FETCHING_NEXT_RELEASE = 102;
+
+    /** @var string */
+    private $apiUrl;
+
+    /** @var Output */
+    private $output;
+
+    public function __construct(string $apiUrl, ?Output $output = null)
+    {
+        $this->apiUrl = $apiUrl;
+        $this->output = $output ?: new StdOutput();
+    }
+
+    public function update(string $currentVersion, string $targetPath)
+    {
+        if (empty($currentVersion)) {
+            throw new NoVersionDeterminedException();
+        }
+
+        $latest = $this->getLatestRelease();
+        if (!$latest->isMoreRecentThan($currentVersion)) {
+            throw new NoUpdateAvailableException($currentVersion);
+        }
+
+        $temp = tempnam(sys_get_temp_dir(), 'workspace-update-') . '.phar';
+
+        try {
+            $this->output->infof('Downloading new version (%s) from %s', $latest->getVersion(), $latest->getUrl());
+            $releaseData = @file_get_contents($latest->getUrl());
```

Comment:
could use the `FileLoader` service instead of `file_get_contents`

Reply:
we could, but what advantage do we get?

Reply:
in other places it was to safely load a file and it (could) can also be mocked out. as we're doing other FS ops here it's not worth it.

Reply:
yeah, in this context it is fairly limited